src/antlr/LineageVisitor.ts

```diff
@@ -38,14 +39,34 @@ export class LineageVisitor<TableData, ColumnData> extends QueryStructureVisitor
     });
   }
 
+  private remapTableIds(newId: string, existingId: string) {
+    this.lineage.forEach(l => {
+      if (l.type === "edge" && l.source.tableId === newId) {
+        l.source.tableId = existingId;
+      }
+    });
+  }
+
   onRelation(relation: TableRelation | QueryRelation, alias?: string): void {
     let label = alias ?? relation.id;
+
+    const columns = relation.columns.map(c => {
+      return {
+        id: c.id,
+        label: c.label,
+        range: c.range,
+        data: c.data as ColumnData,
+        isAssumed: c.isAssumed
+      };
+    });
+
     if (relation instanceof TableRelation) {
       if (this.mergedLeaves) {
         const key = JSON.stringify(relation.tablePrimary);
         const existing = this.usedTables.get(key);
         if (existing !== undefined) {
           this.deduplicateTables.set(relation.id, existing);
+          this.remapTableIds(relation.id, existing);
```

Comment:
Is it for the case when unknown table is used twice in the query?

Reply:
This is meant to fix references to tables that should have been replaced due to `mergedLeaves`. Since tables are reported rather late now, we will often see references to columns before `onRelation` is called, and thus before `deduplicateTables` is populated. The method `remapTableIds` goes through the references and fixes the id's for relations. It's not the prettiest solution, so I am open for suggestions.

Either way, this line definitely needs a comment.

Reply:
I think we can deduplicate tables at later stage when all tables already read. Probably we can just collect tables and edges separately and then "massage" them. Please look at `getLineage()` implementation in `index.ts`. There is already some logic.